monotone.1
.TH "monotone" 1
.SH NAME
monotone \- distributed version control system
.SH SYNOPSIS
\fBmonotone\fP \fI[options] <command> [parameters]\fP
.P
Options, which affect global behavior or set default values, come
first in the argument list. A single command must follow, indicating
the operation to perform, followed by parameters which vary depending
on the command.
.SS Note
This manpage is a summary of some of the features and commands of
\fBmonotone\fP, but it is not the most detailed source of information
available. For a complete discussion of the concepts and a tutorial on
its use, please refer to the texinfo manual (via the \fBinfo
monotone\fP command, or online).
.SS Commands
.TP 
\fBcomment\fP \fI<id>\fP
Write a comment cert for a revision.
.TP
\fBapprove\fP \fI<id>\fP
Make a "branch" cert approving of a revision's membership in a branch.
.TP
\fBdisapprove\fP \fI<id1>\fP
Disapprove of a revision, committing the inverse changes as as a
descendant of the disapproved revision.
.TP
\fBtag\fP \fI<id> <tagname>\fP
Put a symbolic tag cert on a revision.
.TP
\fBtestresult\fP \fI<id> (0|1|true|false|yes|no|pass|fail)\fP
Indicate a passing or failing test result on a revision.
.TP
\fBdiff \fI[--revision=<id1> [--revision=<id2>] ] [<pathname>...]\fP
Show diffs between workspace and database.
.TP
\fBstatus \fI[<pathname>...]\fP
Show status of workspace.
.TP
\fBlog\fP \fI[id] \fP
Show historical log of revisions, starting from workspace
base revision, or \fI[id]\fP if given.
.TP
\fBcert\fP \fI<id> <certname> [certval]\fP
Create a custom cert for a revision. Reads cert value
from stdin if no value given on command line.
.TP
\fBgenkey\fP \fI<keyid>\fP
Generate an RSA key-pair and store it in the database.
.TP
\chkeypass\fP \fI<keyid>\fP
Change passphrase of the private half of a key.
.TP
\fBlist certs\fP \fI<id>\fP
List certs associated with revision.
.TP
\fBlist keys\fP \fI[partial-id]\fP
List keys matching glob, or list all keys if no glob given.
.TP
\fBlist branches\fP
List all branches.
.TP
\fBlist tags\fP
List all tags.
.TP
\fBlist vars \fI[<domain>]\fP
List all vars (possibly limited by domain).
.TP
\fBlist unknown \fI[<pathname...]\fP
List files in workspace, but not in revision's manifest or
work list.
.TP
\fBlist ignored \fI[<pathname...]\fP
List files intentionally ignored due to the ignore_file hook.
.TP
\fBlist missing \fI[<pathname...]\fP
List files in revision's manifest, but not in workspace.
.TP
\fBlist changed \fI[<pathname...]\fP
List files in workspace that have changed compared to the base
revision.
.TP
\fBfdata\fP \fI<id>\fP
Write file data packet to stdout.
.TP
\fBfdelta\fP \fI<oldid> <newid>\fP
Write file delta packet to stdout.
.TP
\fBmdata\fP \fI<id>\fP
Write manifest data packet to stdout.
.TP
\fBmdelta\fP \fI<oldid> <newid>\fP
Write manifest delta packet to stdout.
.TP
\fBcerts\fP \fI<id>\fP
Write revision cert packets to stdout.
.TP
\fBrdata\fP \fI<id>\fP
Write revision data packet to stdout.
.TP
\fBprivkey\fP \fI<id>\fP
Write private key packet to stdout.
.TP
\fBpubkey\fP \fI<id>\fP
Write public key packet to stdout.
.TP
\fBread\fP \fI[<file1> [<file2> [...]]]\fP
Read packets from files or stdin. 
.TP
\fBcvs_import\fP \fI<cvsroot>/<module>\fP
Import all versions in CVS module. Reconstructs revisions and converts
metadata to certificates. A private signing key must already exist in
the database.
.TP
\fBrcs_import\fP \fI<rcsfile> ...\fP
Import all file versions in RCS files. Does not reconstruct revisions
across the entire tree.  You do not want this command, it is for
debugging; use cvs_import.
.TP
\fBcheckout\fP \fI[--revision=revision-id]\fP \fI[<directory>]\fP
Check out revision from database.
.TP
\fBco\fP \fI[--revision=revision-id]\fP \fI[<directory>]\fP
Check out revision from database; an alias for \fBcheckout\fP.
.TP
\fBcat\fP \fI(file|manifest) <id>\fP
Write file or manifest from database to stdout.
.TP
\fBheads\fP
Show unmerged heads of branch, or report when branch is merged.
.TP
\fBmerge\fP
Merge unmerged heads of branch.
.TP
\fBadd\fP \fI<pathname> [...]\fP
Add files to workspace. adding a file does not copy it into the database,
merely adds it to the work list. You must \fBcommit\fP your changes in order
to copy added files to the database.
.TP
\fBdrop\fP \fI<pathname> [...]\fP
Drop files from workspace. Files are not deleted from workspace, 
merely noted as removals in the work list.
.TP
\fBrename\fP \fI<src> \fI<dst>\fP
Rename files from \fI<src> \fP to \fI<dst> \fP in workspace.
.TP
\fBcommit\fP \fI[(--message=<log message>|--message-file=<file>)] [<pathname>...]\fP
Commit workspace to database.  Each commit has a changelog message
associated with it.  If --message is provided on the command line, it
is used; if --message-file is provided, the content of the
named file will be used as a commit message. If the filename is '-'
the commit message will be read from standard input.  Otherwise a log
message editor will be invoked.  If the file MT/log exists
and is non-empty, its content is used to prefill the editor.  You
cannot specify both --message and --message-file at the same time, and
if MT/log exists and is non-empty, you can cannot specify either of them
at all.
.TP
\fBupdate\fP  \fI[revision-id]\fP
Update workspace.
.TP
\fBrefresh_inodeprints\fP 
Turn on inodeprints mode, and force a cache refresh.
.TP
\fBpush\fP \fI[<host> [<glob>]]\fP 
Push contents of \fI<glob>\fP to database on \fI<host>\fP.
.TP
\fBpull\fP \fI[<host> [<glob>]]\fP 
Push contents of \fI<glob>\fP from database on \fI<host>\fP.
.TP
\fBsync\fP \fI<host> <glob>\fP 
Sync contents of \fI<glob>\fP with database on \fI<host>\fP.
.TP
\fBserve\fP \fI[--pid-file=<path>] <host>[:<port> <glob>\fP 
Serve contents of \fI<glob>\fP at network address \fI<host>\fP, on the
port \fI<port>\fP.  If \fI<port>\fP isn't given, 4691 is used. If a
--pid-file option is provided on the command line, monotone will store
the process id of the server in the specified file.
.TP
\fBset\fP \fI<domain> <name> <value>\fP 
Set the db var \fI<name>\fP in domain \fI<domain>\fP to value
\fI<value>\fP. 
.TP
\fBunset\fP \fI<domain> <name>\fP 
Delete any setting for db var \fI<name>\fP in domain \fI<domain>\fP.
.TP
\fBautomate\fP \fI(interface_version|heads|ancestors|attributes|parents|descendents|children|graph|erase_ancestors|toposort|ancestry_difference|leaves|inventory|stdio|certs|select)\fP
Scripting interface.
.TP
\fBdb\fP \fI(init|info|version|dump|load|migrate|rebuild|execute|check)\fP
Manipulate database state.
.SH DESCRIPTION
Monotone is a version control system, which allows you to keep old
versions of files, as well as special \fImanifest files\fP which
describe the location of files in a tree. Unlike other systems,
versions in monotone are \fIidentified\fP by cryptographic hash, and
operations are authenticated by individual users' evaluating
cryptographic signatures on meta-data, rather than any central
authority.

Monotone keeps a collection of versions in a single-file relational
database. It is essentially serverless, using network servers only as
untrusted communication facilities. A monotone database is a regular
file, which contains all the information needed to extract previous
versions of files, verify signatures, merge and modify versions, and
communicate with network servers.
.SH OPTIONS
.TP
\fB--help\fP
Print help message.
.TP
\fB--debug\fP
Turn on debugging log on standard error stream. This is very
verbose. Default is to be silent, unless an error occurs, in which
case failure log is dumped.
.TP
\fB--quiet\fP
Turn off normal progress messages.
.TP
\fB--dump=\fP\fI<file>\fP
Dump debugging log to \fIfile\fP on failure.
.TP
\fB--nostd\fP
Do not evaluate "standard" lua hooks compiled into \fBmonotone\fP.
.TP
\fB--norc\fP
Do not load lua hooks from user's \fB~/.monotonerc\fP file.
.TP
\fB--rcfile=\fP\fI<file>\fP
Load extra lua hooks from \fIfile\fP (may be given multiple times).
.TP
\fB--db=\fP\fI<file>\fP
Use database in \fIfile\fP.
.TP
\fB--key=\fP\fI<keyid>\fP
Use \fIkeyid\fP for operations which produce RSA signatures. Default
is inferred from presence of unique private key in database. Can also
be customized on a per-branch basis with hook function 
\fBget_branch_key(branchname)\fP.
.TP
\fB-k\fP \fI<keyid>\fP
An alias for \fB--key=\fP\fI<keyid>\fP
.TP
\fB--branch=\fP\fI<branchname>\fP
Use \fIbranchname\fP for operations on a branch. Default is inferred
in operations on existing branches (commit, update, etc).
.TP
\fB-b\fP \fI<branchname>\fP
An alias for \fB--branch=\fP\fI<branchname>\fP
.TP
\fB--ticket=\fP\fIdot|count|none\fP
Use the given method to print tickers.  The \fBcount\fP method prints
the count for each ticker on one line, incrementing the numbers in
place, while the \fBdot\fP method prints a continuous string of
characters (like some programs provide a progress line of dots).
\fBnone\fP prints nothing.  The default is \fBcount\fP.
.TP
\fB--revision=\fP\fI<id>\fP
Used to specify one or more revision ids to various commands.
.TP
\fB-r\fP \fI<id>\fP
An alias for \fB--revision=\fP\fI<id>\fP
.TP
\fB--message=\fI<log message>\fP
Use the given message as the changelog when committing a new revision

NEWS
Sat Feb 11 13:32:51 PST 2006

        0.26pre2 release.  Inching towards 0.26.  If you are using
        0.25 or earlier, then make sure to read the very important
        notes for 0.26pre1, below.  In particular, like 0.26pre1, this
        is a pre-release for testing.  Do not package it.  DO NOT USE
        THIS RELEASE UNLESS YOU WANT TO BE A DAREDEVIL.

        (Though, in fact, in a month of usage, only one bug has been
        found in the new history code, and it was both minor and
        harmless.  It has additionally been fixed.)

        Database changes:

        - SQLite 3.3.3 has been imported.  3.3 introduces a new database
          format that is not backwards compatible with earlier 3.x releases.
          New databases will be created using this new format.  Existing
          databases remain compatible, and are not converted automatically.
          Existing databases can be converted by performing a database
          vacuum ('monotone db execute vacuum').

        New features:

        - New hook validate_commit_message -- use to verify that all
          commit messages meet arbitrary user-defined rules.

        UI improvements:

        - rename (and mv) commands now accept a broader range of
          syntax:
            monotone rename foo some_dir
              -> renames foo to some_dir/foo
            monotone rename foo bar baz some_dir
              -> moves foo, bar, and baz to some_dir/foo,
                 some_dir/bar, and some_dir/baz
        - Print a warning if it looks like a user has made a quoting
          mistake on push/pull/sync/serve (windows cmd.exe has
          confusing rules here).
        - New command "ls changed".
        - New option "--next" to log, which displays descendents of
          the start revision.
        - Updating to an arbitrary revision now works again (as it did
          in 0.25 and earlier).  This allows one to, for instance,
          switch a working copy to another head, or back up to an
          earlier version, while preserving uncommitted changes.
        - New option --brief to annotate, gives somewhat more friendly
          output.
        - Fixed bug that made ticker output from netsync inaccurate.
        - In 'log', --no-merges is now the default, use --merges to
          override.
        - If the database is in the working copy, then it is always
          ignored.

        Bugs:

        - 'serve' with no --bind should now work on systems where the
          C library has IPv6 support, but the kernel does not.
        - Compile fixes for GCC 4.1 pre-releases.

        Other:
        - Better detection when users have not run "rosterify", and
          more helpful suggestions on what to do in this case.
        - Documentation, translation, error message,
          etc. improvements.
        - Updates to contrib/mtbrowse.sh, simple shell-based monotone
          interface.
        - Updates to many other contrib/ files, mostly to maintain
          compatibility with monotone changes.

Sun Jan  8 01:08:56 PST 2006

        0.26pre1 release.  Massive rewrites, released for shakedown.
        This release is also dedicated to Shweta Narayan.

        This release includes massive changes compared to 0.25.  The
        core versioning code has all been replaced with a completely
        different mechanism.  Data formats and the netsync protocol
        have changed in incompatible ways.

        Migration to 0.26pre1 or later is irreversible and requires a
        flag day for your project.  See UPGRADE for details.  Note
        that we DO NOT recommend upgrading at this time; see below.

        If you have been following the development list for the last
        few months, you may have heard about "rosters" -- this is the
        name for the new core data structure we use.  While the code
        is completely different, the user experience should not be
        very different.  You will never see a roster, unless you are
        debugging monotone itself; everything still revolves around
        revisions, manifests, and certs.

        While this new code has extensive tests, because of these
        incompatibilities, it has never been used for real work.  The
        purpose of this release is to make a version available for the
        monotone developers to begin using for day-to-day work, to
        shake out bugs.

        Let's say that again in caps: THIS CODE IS PROBABLY BUGGY, DO
        NOT USE IT IN PRODUCTION UNLESS YOU WANT TO BE A DAREDEVIL.
        

monotone.cc
// -*- mode: C++; c-file-style: "gnu"; indent-tabs-mode: nil -*-
// copyright (C) 2002, 2003 graydon hoare <graydon@pobox.com>
// all rights reserved.
// licensed to the public under the terms of the GNU GPL (>= 2)
// see the file COPYING for details

#include "config.h"

#include "popt/popt.h"
#include <cstdio>
#include <strings.h>
#include <iterator>
#include <iostream>
#include <fstream>
#include <sstream>
#include <locale.h>

#include <stdlib.h>

#include "botan/botan.h"

#include "i18n.h"

#include "app_state.hh"
#include "commands.hh"
#include "sanity.hh"
#include "cleanup.hh"
#include "file_io.hh"
#include "transforms.hh"
#include "ui.hh"
#include "mt_version.hh"
#include "options.hh"
#include "paths.hh"

// main option processing and exception handling code

using namespace std;

char * argstr = NULL;
long arglong = 0;

// Options are split between two tables.  The first one is command-specific
// options (hence the `c' in `coptions').  The second is the global one
// with options that aren't tied to specific commands.
//
// the intent is to ensure that any command specific options mean the same
// thing to all commands that use them

struct poptOption coptions[] =
  {
    {"branch", 'b', POPT_ARG_STRING, &argstr, OPT_BRANCH_NAME, gettext_noop("select branch cert for operation"), NULL},
    {"revision", 'r', POPT_ARG_STRING, &argstr, OPT_REVISION, gettext_noop("select revision id for operation"), NULL},
    {"message", 'm', POPT_ARG_STRING, &argstr, OPT_MESSAGE, gettext_noop("set commit changelog message"), NULL},
    {"message-file", 0, POPT_ARG_STRING, &argstr, OPT_MSGFILE, gettext_noop("set filename containing commit changelog message"), NULL},
    {"date", 0, POPT_ARG_STRING, &argstr, OPT_DATE, gettext_noop("override date/time for commit"), NULL},
    {"author", 0, POPT_ARG_STRING, &argstr, OPT_AUTHOR, gettext_noop("override author for commit"), NULL},
    {"depth", 0, POPT_ARG_LONG, &arglong, OPT_DEPTH, gettext_noop("limit the number of levels of directories to descend"), NULL},
    {"last", 0, POPT_ARG_LONG, &arglong, OPT_LAST, gettext_noop("limit log output to the last number of entries"), NULL},
    {"next", 0, POPT_ARG_LONG, &arglong, OPT_NEXT, gettext_noop("limit log output to the next number of entries"), NULL},
    {"pid-file", 0, POPT_ARG_STRING, &argstr, OPT_PIDFILE, gettext_noop("record process id of server"), NULL},
    {"brief", 0, POPT_ARG_NONE, NULL, OPT_BRIEF, gettext_noop("print a brief version of the normal output"), NULL},
    {"diffs", 0, POPT_ARG_NONE, NULL, OPT_DIFFS, gettext_noop("print diffs along with logs"), NULL},
    {"no-merges", 0, POPT_ARG_NONE, NULL, OPT_NO_MERGES, gettext_noop("exclude merges when printing logs"), NULL},
    {"set-default", 0, POPT_ARG_NONE, NULL, OPT_SET_DEFAULT, gettext_noop("use the current arguments as the future default"), NULL},
    {"exclude", 0, POPT_ARG_STRING, &argstr, OPT_EXCLUDE, gettext_noop("leave out anything described by its argument"), NULL},
    {"unified", 0, POPT_ARG_NONE, NULL, OPT_UNIFIED_DIFF, gettext_noop("use unified diff format"), NULL},
    {"context", 0, POPT_ARG_NONE, NULL, OPT_CONTEXT_DIFF, gettext_noop("use context diff format"), NULL},
    {"external", 0, POPT_ARG_NONE, NULL, OPT_EXTERNAL_DIFF, gettext_noop("use external diff hook for generating diffs"), NULL},
    {"diff-args", 0, POPT_ARG_STRING, &argstr, OPT_EXTERNAL_DIFF_ARGS, gettext_noop("argument to pass external diff hook"), NULL},
    {"lca", 0, POPT_ARG_NONE, NULL, OPT_LCA, gettext_noop("use least common ancestor as ancestor for merge"), NULL},
    {"execute", 'e', POPT_ARG_NONE, NULL, OPT_EXECUTE, gettext_noop("perform the associated file operation"), NULL},
    {"bind", 0, POPT_ARG_STRING, &argstr, OPT_BIND, gettext_noop("address:port to listen on (default :4691)"), NULL},
    {"missing", 0, POPT_ARG_NONE, NULL, OPT_MISSING, gettext_noop("perform the operations for files missing from workspace"), NULL},
    {"unknown", 0, POPT_ARG_NONE, NULL, OPT_UNKNOWN, gettext_noop("perform the operations for unknown files from workspace"), NULL},
    {"key-to-push", 0, POPT_ARG_STRING, &argstr, OPT_KEY_TO_PUSH, gettext_noop("push the specified key even if it hasn't signed anything"), NULL},
    {"drop-attr", 0, POPT_ARG_STRING, &argstr, OPT_DROP_ATTR, gettext_noop("when rosterifying, drop attrs entries with the given key"), NULL},
    {"no-files", 0, POPT_ARG_NONE, NULL, OPT_NO_FILES, gettext_noop("exclude files when printing logs"), NULL},
    {"recursive", 'R', POPT_ARG_NONE, NULL, OPT_RECURSIVE, gettext_noop("also operate on the contents of any listed directories"), NULL},
    { NULL, 0, 0, NULL, 0, NULL, NULL }
  };

struct poptOption options[] =
  {
    // Use the coptions table as well.
    { NULL, 0, POPT_ARG_INCLUDE_TABLE, coptions, 0, NULL, NULL },

    {"debug", 0, POPT_ARG_NONE, NULL, OPT_DEBUG, gettext_noop("print debug log to stderr while running"), NULL},
    {"dump", 0, POPT_ARG_STRING, &argstr, OPT_DUMP, gettext_noop("file to dump debugging log to, on failure"), NULL},
    {"log", 0, POPT_ARG_STRING, &argstr, OPT_LOG, gettext_noop("file to write the log to"), NULL},
    {"quiet", 0, POPT_ARG_NONE, NULL, OPT_QUIET, gettext_noop("suppress log and progress messages"), NULL},
    {"help", 'h', POPT_ARG_NONE, NULL, OPT_HELP, gettext_noop("display help message"), NULL},
    {"version", 0, POPT_ARG_NONE, NULL, OPT_VERSION, gettext_noop("print version number, then exit"), NULL},
    {"full-version", 0, POPT_ARG_NONE, NULL, OPT_FULL_VERSION, gettext_noop("print detailed version number, then exit"), NULL},
    {"xargs", '@', POPT_ARG_STRING, &argstr, OPT_ARGFILE, gettext_noop("insert command line arguments taken from the given file"), NULL},
    {"ticker", 0, POPT_ARG_STRING, &argstr, OPT_TICKER, gettext_noop("set ticker style (count|dot|none)"), NULL},
    {"nostd", 0, POPT_ARG_NONE, NULL, OPT_NOSTD, gettext_noop("do not load standard lua hooks"), NULL},
    {"norc", 0, POPT_ARG_NONE, NULL, OPT_NORC, gettext_noop("do not load ~/.monotone/monotonerc or MT/monotonerc lua files"), NULL},
    {"rcfile", 0, POPT_ARG_STRING, &argstr, OPT_RCFILE, gettext_noop("load extra rc file"), NULL},
    {"key", 'k', POPT_ARG_STRING, &argstr, OPT_KEY_NAME, gettext_noop("set key for signatures"), NULL},
    {"db", 'd', POPT_ARG_STRING, &argstr, OPT_DB_NAME, gettext_noop("set name of database"), NULL},
    {"root", 0, POPT_ARG_STRING, &argstr, OPT_ROOT, gettext_noop("limit search for workspace to specified root"), NULL},
    {"verbose", 0, POPT_ARG_NONE, NULL, OPT_VERBOSE, gettext_noop("verbose completion output"), NULL},
    {"keydir", 0, POPT_ARG_STRING, &argstr, OPT_KEY_DIR, gettext_noop("set location of key store"), NULL},
    {"confdir", 0, POPT_ARG_STRING, &argstr, OPT_CONF_DIR, gettext_noop("set location of configuration directory"), NULL},
    { NULL, 0, 0, NULL, 0, NULL, NULL }
  };

// there are 3 variables which serve as roots for our system.
//
// "global_sanity" is a global object, which contains the error logging
// system, which is constructed once and used by any nana logging actions.
// see cleanup.hh for it
//
// "cmds" is a static table in commands.cc which associates top-level
// commands, given on the command-line, to various version control tasks.
//
// "app_state" is a non-static object type which contains all the
// application state (filesystem, database, network, lua interpreter,
// etc). you can make more than one of these, and feed them to a command in
// the command table.

// our main function is run inside a boost execution monitor. this monitor
// portably sets up handlers for various fatal conditions (signals, win32
// structured exceptions, etc) and provides a uniform reporting interface
// to any exceptions it catches. we augment this with a helper atexit()
// which will also dump our internal logs when an explicit clean shutdown
// flag is not set.
//
// in other words, this program should *never* unexpectedly terminate
// without dumping some diagnostics.

void 
dumper() 
{
  if (!global_sanity.clean_shutdown)
    global_sanity.dump_buffer();
  
  Botan::Init::deinitialize();
}


struct 
utf8_argv
{
  int argc;
  char **argv;

  explicit utf8_argv(int ac, char **av)
    : argc(ac),
      argv(static_cast<char **>(malloc(ac * sizeof(char *))))
  {
    I(argv != NULL);
    for (int i = 0; i < argc; ++i)
      {
        external ext(av[i]);
        utf8 utf;
        system_to_utf8(ext, utf);
        argv[i] = static_cast<char *>(malloc(utf().size() + 1));
        I(argv[i] != NULL);
        memcpy(argv[i], utf().data(), utf().size());
        argv[i][utf().size()] = static_cast<char>(0);
    }
  }

  ~utf8_argv() 
  {
    if (argv != NULL)
      {
        for (int i = 0; i < argc; ++i)
          if (argv[i] != NULL)
            free(argv[i]);
        free(argv);
      }    
  }
};

// Stupid type system tricks: to use a cleanup_ptr, we need to know the return
// type of the cleanup function.  But popt silently changed the return type of
// poptFreeContext at some point, I guess because they thought it would be
// "backwards compatible".  We don't actually _use_ the return value of
// poptFreeContext, so this little wrapper works.
static void
my_poptFreeContext(poptContext con)
{
  poptFreeContext(con);
}

// Read arguments from a file.  The special file '-' means stdin.
// Returned value must be free()'d, after arg parsing has completed.
static void
my_poptStuffArgFile(poptContext con, utf8 const & filename)
{
  utf8 argstr;
  {
    data dat;
    read_data_for_command_line(filename, dat);
    external ext(dat());
    system_to_utf8(ext, argstr);
  }

  const char **argv = 0;
  int argc = 0;
  int rc;

  // Parse the string.  It's OK if there are no arguments.
  rc = poptParseArgvString(argstr().c_str(), &argc, &argv);
  N(rc >= 0 || rc == POPT_ERROR_NOARG,
    F("problem parsing arguments from file %s: %s")
    % filename % poptStrerror(rc));

  if (rc != POPT_ERROR_NOARG)
    {
      // poptStuffArgs does not take an argc argument, but rather requires that
      // the argv array be null-terminated.
      I(argv[argc] == NULL);
      N((rc = poptStuffArgs(con, argv)) >= 0,
        F("weird error when stuffing arguments read from %s: %s\n")
        % filename % poptStrerror(rc));
    }

  free(argv);
}

static string
coption_string(int o)
{
  char buf[2] = { 0,0 };
  for(struct poptOption *opt = coptions; opt->val; opt++)
    if (o == opt->val)
      {
        buf[0] = opt->shortName;
        return opt->longName
          ? string("--") + string(opt->longName)
          : string("-") + string(buf);
      }
  return string();
}

int 
cpp_main(int argc, char ** argv)
{
  int ret = 0;

  atexit(&dumper);

  // go-go gadget i18n

  setlocale(LC_ALL, "");
  bindtextdomain(PACKAGE, LOCALEDIR);
  textdomain(PACKAGE);


  // we want to catch any early informative_failures due to charset
  // conversion etc
  try
  {

  {
    std::ostringstream cmdline_ss;
    for (int i = 0; i < argc; ++i)
      {
        if (i)
          cmdline_ss << ", ";
        cmdline_ss << "'" << argv[i] << "'";
      }
    L(FL("command line: %s\n") % cmdline_ss.str());
  }

  L(FL("set locale: LC_ALL=%s\n")
    % (setlocale(LC_ALL, NULL) == NULL ? "n/a" : setlocale(LC_ALL, NULL)));

  // Set up secure memory allocation etc
  Botan::Init::initialize();
  Botan::set_default_allocator("malloc");
  
  // decode all argv values into a UTF-8 array

  save_initial_path();
  utf8_argv uv(argc, argv);

  string prog_name(uv.argv[0]);

  // prepare for arg parsing

  cleanup_ptr<poptContext, void> 
    ctx(poptGetContext(NULL, argc, (char const **) uv.argv, options, 0),
        &my_poptFreeContext);

  set<int> local_options;
  for (poptOption *opt = coptions; opt->val; opt++)
    local_options.insert(opt->val);

  // process main program options

  int opt;
  bool requested_help = false;
  set<int> used_local_options;

  poptSetOtherOptionHelp(ctx(), _("[OPTION...] command [ARGS...]\n"));

  try
    {
      app_state app;

      app.prog_name = prog_name;

      while ((opt = poptGetNextOpt(ctx())) > 0)
        {
          if (local_options.find(opt) != local_options.end())
            used_local_options.insert(opt);

          switch(opt)
            {
            case OPT_DEBUG:
              global_sanity.set_debug();
              break;

            case OPT_QUIET:
              global_sanity.set_quiet();
              break;

            case OPT_NOSTD:
              app.set_stdhooks(false);
              break;

            case OPT_NORC:
              app.set_rcfiles(false);
              break;

            case OPT_VERBOSE:
              app.set_verbose(true);
              break;

            case OPT_RCFILE:
              app.add_rcfile(string(argstr));
              break;

            case OPT_DUMP:
              global_sanity.filename = system_path(argstr);
              break;

            case OPT_LOG:
              ui.redirect_log_to(system_path(argstr));
              break;

            case OPT_DB_NAME:
              app.set_database(system_path(argstr));
              break;

            case OPT_KEY_DIR:
              app.set_key_dir(system_path(argstr));
              break;

            case OPT_CONF_DIR:
              app.set_confdir(system_path(argstr));
              break;

            case OPT_TICKER:
              if (string(argstr) == "dot")
                ui.set_tick_writer(new tick_write_dot);
              else if (string(argstr) == "count")
                ui.set_tick_writer(new tick_write_count);
              else if (string(argstr) == "none")
                ui.set_tick_writer(new tick_write_nothing);
              else
                requested_help = true;
              break;

            case OPT_KEY_NAME:
              app.set_signing_key(string(argstr));
              break;

            case OPT_BRANCH_NAME:
              app.set_branch(string(argstr));
              app.set_is_explicit_option(OPT_BRANCH_NAME);
              break;

            case OPT_VERSION:
              print_version();
              global_sanity.clean_shutdown = true;
              return 0;

            case OPT_FULL_VERSION:
              print_full_version();
              global_sanity.clean_shutdown = true;
              return 0;

            case OPT_REVISION:
              app.add_revision(string(argstr));
              break;

            case OPT_MESSAGE:
              app.set_message(string(argstr));
              app.set_is_explicit_option(OPT_MESSAGE);
              break;

            case OPT_MSGFILE:
              app.set_message_file(string(argstr));
              app.set_is_explicit_option(OPT_MSGFILE);
              break;

            case OPT_DATE:
              app.set_date(string(argstr));
              break;

            case OPT_AUTHOR:
              app.set_author(string(argstr));
              break;

            case OPT_ROOT:
              app.set_root(system_path(argstr));
              break;

            case OPT_LAST:
              app.set_last(arglong);
              break;

            case OPT_NEXT:
              app.set_next(arglong);
              break;

            case OPT_DEPTH:
              app.set_depth(arglong);
              break;

            case OPT_BRIEF:
              global_sanity.set_brief();
              break;

            case OPT_DIFFS:
              app.diffs = true;
              break;

            case OPT_NO_MERGES:
              app.no_merges = true;
              break;

            case OPT_SET_DEFAULT:
              app.set_default = true;
              break;

            case OPT_EXCLUDE:
              app.add_exclude(utf8(string(argstr)));
              break;

            case OPT_PIDFILE:
              app.set_pidfile(system_path(argstr));
              break;

            case OPT_ARGFILE:
              my_poptStuffArgFile(ctx(), utf8(string(argstr)));
              break;

            case OPT_UNIFIED_DIFF:
              app.set_diff_format(unified_diff);
              break;

            case OPT_CONTEXT_DIFF:
              app.set_diff_format(context_diff);
              break;

            case OPT_EXTERNAL_DIFF:
              app.set_diff_format(external_diff);
              break;
              
            case OPT_EXTERNAL_DIFF_ARGS:
              app.set_diff_args(utf8(string(argstr)));
              break;

            case OPT_LCA:
              app.use_lca = true;
              break;

            case OPT_EXECUTE:
              app.execute = true;
              break;

            case OPT_BIND:
              {
                std::string arg(argstr);
                std::string addr_part, port_part;
                size_t l_colon = arg.find(':');
                size_t r_colon = arg.rfind(':');
                
                // not an ipv6 address, as that would have at least two colons
                if (l_colon == r_colon)
                  {
                    addr_part = (r_colon == std::string::npos ? arg : arg.substr(0, r_colon));
                    port_part = (r_colon == std::string::npos ? "" :  arg.substr(r_colon+1, arg.size() - r_colon));
                  }
                else
                  { 
                    // IPv6 addresses have a port specified in the style: [2001:388:0:13::]:80
                    size_t squareb = arg.rfind(']');
                    if ((arg.find('[') == 0) && (squareb != std::string::npos))
                      {
                        if (squareb < r_colon)
                          port_part = (r_colon == std::string::npos ? "" :  arg.substr(r_colon+1, arg.size() - r_colon));
                        else
                          port_part = "";
                        addr_part = (squareb == std::string::npos ? arg.substr(1, arg.size()) : arg.substr(1, squareb-1));
                      }
                    else 
                      {
                        addr_part = arg;
                        port_part = "";
                      }
                  }
                app.bind_address = utf8(addr_part);
                app.bind_port = utf8(port_part);
              }
              app.set_is_explicit_option(OPT_BIND);
              break;

            case OPT_MISSING:
              app.missing = true;
              break;

            case OPT_UNKNOWN:
              app.unknown = true;
              break;

            case OPT_KEY_TO_PUSH:
              {
                app.add_key_to_push(string(argstr));
              }
              break;

            case OPT_DROP_ATTR:
              app.attrs_to_drop.insert(string(argstr));
              break;

            case OPT_NO_FILES:
              app.no_files = true;
              break;

            case OPT_RECURSIVE:
              app.set_recursive();
              break;

            case OPT_HELP:
            default:
              requested_help = true;
              break;
            }
        }

      // verify that there are no errors in the command line

      N(opt == -1,
        F("syntax error near the \"%s\" option: %s") %
          poptBadOption(ctx(), POPT_BADOPTION_NOALIAS) % poptStrerror(opt));

      // complete the command if necessary

      string cmd;
      if (poptPeekArg(ctx()))
        {
          cmd = commands::complete_command(poptGetArg(ctx()));
        }

      // stop here if they asked for help

      if (requested_help)
        {
          throw usage(cmd);     // cmd may be empty, and that's fine.
        }

      // at this point we allow a workspace (meaning search for it
      // and if found read MT/options, but don't use the data quite
      // yet, and read all the monotonercs).  Processing the data
      // from MT/options happens later.
      // Certain commands may subsequently require a workspace or fail
      // if we didn't find one at this point.

      app.allow_workspace();

      // main options processed, now invoke the 
      // sub-command w/ remaining args

      if (cmd.empty())
        {
          throw usage("");
        }
      else
        {
          // Make sure the local options used are really used by the
          // given command.
          set<int> command_options = commands::command_options(cmd);
          for (set<int>::const_iterator i = used_local_options.begin();
               i != used_local_options.end(); ++i)
            N(command_options.find(*i) != command_options.end(),
              F("monotone %s doesn't use the option %s")
              % cmd % coption_string(*i));

          vector<utf8> args;
          while(poptPeekArg(ctx()))
            {
              args.push_back(utf8(string(poptGetArg(ctx()))));
            }
          ret = commands::process(app, cmd, args);
        }
    }
  catch (usage & u)
    {
      // Make sure to hide documentation that's not part of
      // the current command.
      set<int> command_options = commands::command_options(u.which);
      int count = 0;
      for (poptOption *o = coptions; o->val != 0; o++)
        {
          if (command_options.find(o->val) != command_options.end())
            {
              o->argInfo &= ~POPT_ARGFLAG_DOC_HIDDEN;
              L(FL("Removed 'hidden' from option # %d\n") % o->argInfo);
              count++;
            }
          else
            {
              o->argInfo |= POPT_ARGFLAG_DOC_HIDDEN;
              L(FL("Added 'hidden' to option # %d\n") % o->argInfo);
            }
        }
      free((void *)options[0].descrip); options[0].descrip = NULL;
      if (count != 0)
        {
          ostringstream sstr;
          sstr << F("Options specific to '%s %s':") 
            % prog_name % u.which;
          options[0].descrip = strdup(sstr.str().c_str());

          options[0].argInfo |= POPT_ARGFLAG_DOC_HIDDEN;
          L(FL("Added 'hidden' to option # %d\n") % options[0].argInfo);
        }

      poptPrintHelp(ctx(), stdout, 0);
      cout << endl;
      commands::explain_usage(u.which, cout);
      global_sanity.clean_shutdown = true;
      return 2;
    }
  }
  catch (informative_failure & inf)
  {
    ui.inform(inf.what);
    global_sanity.clean_shutdown = true;
    return 1;
  }
  catch (std::ios_base::failure const & ex)
  {
    global_sanity.clean_shutdown = true;
    return 1;
  }

  global_sanity.clean_shutdown = true;
  return ret;
}
